include/utilities/bmi_utilities.hpp

```diff
@@ -0,0 +1,110 @@
+            std::vector<T> result;
+
+            /*
+            * Allows the std::vector constructor to type cast the values as it copies them.
+            * I don't see any other way around the typing issue other than an explicit copy of each...
+            * Now there is an early optimization that allows types that align to pass through uncopied
+            * but that also only works if GetValuePtr returns a compatible pointer that can iterate
+            * on the recieving side correctly.  This may be tricky for certain langague adapters (Fortran?)
+            * Untill this becomes burdensome on memory/time, I suggest copying and converting each value
+            */
+
+            if (type == "long double"){
+                result = helper::make_vector<T>( (long double*) data, num_items);
+            }
+            else if (type == "double"){
+                result = helper::make_vector<T>( (double*) data, num_items);
+            }
+            else if (type == "float"){
+                result = helper::make_vector<T>( (float*) data, num_items);
+            }
+            else if (type == "short" || type == "short int" || type == "signed short" || type == "signed short int"){
+                result = helper::make_vector<T>( (short*) data, num_items);
+            }
+            else if (type == "unsigned short" || type == "unsigned short int"){
+                result = helper::make_vector<T>( (unsigned short*) data, num_items);
+            }
+            else if (type == "int" || type == "signed" || type == "signed int"){
+                result = helper::make_vector<T>( (int*) data, num_items);
+            }
+            else if (type == "unsigned" || type == "unsigned int"){
+                result = helper::make_vector<T>( (unsigned int*) data, num_items);
+            }
+            else if (type == "long" || type == "long int" || type == "signed long" || type == "signed long int"){
+                result = helper::make_vector<T>( (long*) data, num_items);
+            }
+            else if (type == "unsigned long" || type == "unsigned long int"){
+                result = helper::make_vector<T>( (unsigned long*) data, num_items);
+            }
+            else if (type == "long long" || type == "long long int" || type == "signed long long" || type == "signed long long int"){
+                result = helper::make_vector<T>( (long long*) data, num_items);
+            }
+            else if (type == "unsigned long long" || type == "unsigned long long int"){
+                result = helper::make_vector<T>( (unsigned long long*) data, num_items);
+            }
+            else{
+                throw std::runtime_error("Unable to get value of variable " + name +
+                                " as " + boost::typeindex::type_id<T>().pretty_name() + ": no logic for converting variable type " + type);
+            }
+            //clean up the temporary data
+            ::operator delete(data); 
```

Comment:
At this position we will get a memory leak on the above exception.

Reply:
Pretty sure this is a fatal error at this point, but in case it is caught elsewhere, we could `delete` before we `throw` as well.

Reply:
I would add the delete[] before the throw, no reason to leave a potential leak, even on something assumed to be fatal, if we dont need to.